File: Problems/98.validate-binary-search-tree.go
// title: Validate Binary Search Tree
// timestamp: 2024-04-14 20:24:33
// problemUrl: https://leetcode.com/problems/validate-binary-search-tree/
// memory: 5.4 MB
// runtime: 0 ms

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
    valid, _, _ := validBST(root)
    return valid
}

func validBST(root *TreeNode) (valid bool, min int, max int) {
    min, max = 2147483647, -2147483648

    if root == nil {
        return true, min, max
    }
    
    if root.Left != nil {
        if root.Left.Val >= root.Val {
            return false, min, max
        }
        valid, minL, maxL := validBST(root.Left)
        if !valid || maxL >= root.Val {
            return false, min, max
        }
        if min > minL {
            min = minL
        }
        if max < maxL {
            max = maxL
        }
    } else {
        min, max = root.Val, root.Val
    }

    if root.Right != nil {
        if root.Val >= root.Right.Val {
            return false, min, max
        }
        valid, minR, maxR := validBST(root.Right)
        if !valid || minR <= root.Val {
            return false, min, max
        }
        if min > minR {
            min = minR
        }
        if max < maxR {
            max = maxR
        }
    } else {
        max = root.Val
    }

    return true, min, max
}

File: Problems/98.validate-binary-search-tree.py
# title: Validate Binary Search Tree
# timestamp: 2024-12-21 10:33:23
# problemUrl: https://leetcode.com/problems/validate-binary-search-tree/
# memory: 19.8 MB
# runtime: 4 ms

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        def helper(root):
            if not root.left and not root.right:
                return True, root.val, root.val
            
            mi, ma = root.val, root.val
            if root.left:
                res, minL, maxL = helper(root.left)
                if not res or root.val <= maxL:
                    return False, -1, -1
                mi = minL
            if root.right:
                res, minR, maxR = helper(root.right)
                if not res or minR <= root.val:
                    return False, -1, -1
                ma = maxR
            
            return True, mi, ma
        
        res, _, _ = helper(root)
        return res
